API v0.0.3: DICOM download endpoint

MounaSafiHarab opened the pull request:
Added a DICOM download endpoint in v0.0.3 of the Loris Rest API (Redmine 12851).
Documentation adapted for earlier scans (inserted as all series in one tar file), and for newer scans (as of Loris-MRI v17.2) where each series will be handled as a single tar.

MounaSafiHarab commented:
Needs to be rebased once 17.1 is released to include recent changes brought to API v0.0.2

driusan commented on docs/API/LorisRESTAPI.md:
```diff
@@ -696,3 +696,76 @@ The JSON object is of the form:
+A GET request to `/candidates/$CandID/$Visit/dicoms` will return a JSON object of
+all the images which have been acquired for that visit. It will return an object of
+the form:
+
```
I think this should be "one of the following forms", either that or there's something missing here.

MounaSafiHarab replied:
the before 17.1 and starting 17.2 release statements were not intended to be kept at all; so removed those (now we have only one form).

driusan commented on docs/API/LorisRESTAPI.md:
```diff
@@ -696,3 +696,76 @@ The JSON object is of the form:
+# 5.0 DICOM Data
+
+Like the imaging data, the DICOM data mostly lives in the `/candidates/$CandID/$Visit` 
+portion of the REST API namespaces, but is defined in a separate section of this 
```
namespace is singular.

MounaSafiHarab replied:
fixed; I also fixed it in an earlier line as well (line 440)

driusan commented on docs/API/LorisRESTAPI.md:
````diff
@@ -696,3 +696,76 @@ The JSON object is of the form:
+GET /candidates/$CandID/$VisitLabel/dicoms/$Tarname
+```
+
+Returns a ```tar``` file which contains a ```.meta``` and a ```.log``` text files, 
````
"```" is for block quotes in markdown, "`" is for inline things.

MounaSafiHarab replied:
done

driusan commented on docs/API/LorisRESTAPI.md:
````diff
@@ -696,3 +696,76 @@ The JSON object is of the form:
+and a ```.tar.gz``` of the raw DICOM data as acquired during the candidate s
+canning session, and as retrieved from `/candidates/$CandID/$Visit/dicoms`.
+
+Only `GET` is currently supported, but future versions of this API may include `PUT`
````
I'd just say may include other HTTP methods and leave it open for which ones, or remove the "but future versions" part entirely..

MounaSafiHarab replied:
Okay, removed the last part altogether.

MounaSafiHarab commented:
@xlecours 
not sure why it dismissed your review (I did not do it although github says so!)

@driusan 
1) addressed your feedback on the readme; 
2) but in one of the imaging meeting a while back; you suggested I change the way the DicomTars are organized (instead of group concatenating the series; you wanted them to be under an object like SeriesInfo which has an array of the series); so I am about to merge a pull request by @davidblader to this branch, so please take a look at the last 2 commits to see the difference (and if you like last commit, then I will re-update the Readme to reflect the new format).

driusan commented:
@MounaSafiHarab I don't see the change.. It still seems to be a comma separated string list?

MounaSafiHarab commented:
> @MounaSafiHarab I don't see the change.. It still seems to be a comma separated string list?

I did not update the Readme yet; I thought I would if you see the code changes. But I will update the readme as well; this way it is easier to see the text changes

MounaSafiHarab commented on docs/API/LorisRESTAPI.md:
```diff
@@ -709,22 +709,40 @@ GET /candidates/$CandID/$Visit/dicoms
+            "EchoTime" : "30",
+            "SeriesUID" : "3.4.5.6.1507",
+            }],
+        "Tarname" : "DCM_yyyy-mm-dd_ImagingUpload-hh-mm-def456.tar",
```
@driusan 
readme updated to reflect the code changes.
is it better if I do not show the case where the DicomTars is an array of `tars`? this happens (rarely) when a candidate gets asked to repeat a certain sequence (MPRAGE in the example above) in which case we will have 2 .tars for the same visit/session.
I debated if it is better to keep this case in the example, and if so, whether the text just above the `js` block should explain this situation?

driusan commented:
@MounaSafiHarab #3296 is merged, this can be rebased now.

MounaSafiHarab commented:
@driusan 
rebased;
removed the passed manual testing tag as I believe this needs re-tetsing to make sure my rebasing is fine.

driusan commented on docs/API/LorisRESTAPI_v0.0.3.md:
```diff
@@ -696,3 +696,63 @@ The JSON object is of the form:
+
+A GET request to `/candidates/$CandID/$Visit/dicoms` will return a JSON object of
+all the raw DICOM data which have been acquired for that visit. It will return an 
+object of the form:
```
would it be a good idea to include an upload date and an optional 'since' parameter to filter it, similarly to what @xlecours did for the images endpoint?

MounaSafiHarab replied:
yes, good point. But should it be done in this pull request or in a separate one?

MounaSafiHarab replied:
> would it be a good idea to include an upload date and an optional 'since' parameter to filter it, similarly to what @xlecours did for the images endpoint?

good point. I planned to do it in a separate pull request though. Did you want it here?

driusan replied:
Sorry for the delayed response.. no, that's fine if you're planning on doing it separately.

driusan commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,149 @@
+     */
+    public function __construct($method, $CandID, $VisitLabel, $Tarname)
+    {
+        ob_start();
```
why is this here? I'm not sure what it's trying to do, but it seems a little convoluted.

MounaSafiHarab replied:
was there in images/image.php, and the dicom here is clearly a copy of the image one! I tested the download after removing this line (as well as Line 73 below); seems to be fine, so will delete those lines.

driusan commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+            ]
+        );
+        if ($cand_info['Entity_type'] == 'Scanner') {
+            $ID            = "'" . $this->VisitLabel . "'" ;
```
This is vulnerable to an SQL injection attack

MounaSafiHarab replied:
I updated the pull request

MounaSafiHarab commented:
@xlecours removed the passed manal test as there has been a couple of commits since then due to Dave's feedback.

ridz1208 commented:
@driusan please dismiss/re-review

ridz1208 commented:
@xlecours please test/review

driusan commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+        if ($fp !== false) {
+            $this->Header("Content-Type: application/x-tar");
+            $this->Header('Content-Length: '.filesize($fullDir));
+            $this->Header('Content-Disposition: filename='.$this->Tarname);
```
Missing "attachment" in the header.

cmadjar replied:
@MounaSafiHarab for point 2. I agree, it would make more sense to label it with the CandID_PSCID_VL. The only question, do we want to keep a trace of what was the tarname? If so, then maybe it could be downloaded as the tarname appended to the CandID_PSCID_VL? If not, simply CandID_PSCID_VL would work.

Hope that makes sense!

driusan replied:
Do the minc files have a `Content-Disposition` header?

The values for Content-Disposition headers are "attachment" or "inline" for determining if the URL is intended to be downloaded or displayed inline in a page. "filename" is an argument for attachment (since it doesn't make sense for inline..), which is why I say it's missing "attachment".. filename by itself isn't a valid value for the headers without specifying what the disposition is.

I think the filename should be the same as the URL endpoint, whatever that is.

davidblader replied:
sorry i have another dumb question but what's the difference between doing `$this->Header(...)` and `$this->header(...)`
is `Header()` just a wrapper?

MounaSafiHarab replied:
@davidblader ; I am assuming the question is addressed to @driusan , as I do not know (the changes all come from the Image

@driusan @cmadjar 
if the filename is to be kept as the URL, then it is the name I currently have 
@cmadjar if we were to download files with a different filename, it should come in a different pull request, simply because the changes to `get_file.php` will impact many other places as well (DICOM archive, Media, Imaging Browser potentially)... but I like your suggestion of appending as opposed to changing the original filename; might be the best compromise?

MounaSafiHarab replied:
> Do the minc files have a Content-Disposition header?

yes they do. A big chunk of this script comes from the MINC counterpart.

driusan replied:
@davidblader Function names in `PHP` are case insensitive. There's no difference between the two lines you pasted.

davidblader commented on docs/API/LorisRESTAPI_v0.0.3.md:
```diff
@@ -728,3 +728,63 @@ The JSON object is of the form:
+
+Returns a `tar` file which contains a `.meta` and a `.log` text files, 
+and a `.tar.gz` of the raw DICOM data as acquired during the candidate
+scanning session, and as retrieved from `/candidates/$CandID/$Visit/dicoms`.
```
might be useful to specify that the raw data should be `>`'d into a file on the user's / client's system? depends on how this plans to be used

MounaSafiHarab replied:
I changed the word `returns` to `downloads` as this is what this will do.

davidblader commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+            $ID            = ":PVL";
+            $params['PVL'] = $this->VisitLabel;
+        } else {
+            $ID = "LOWER(CONCAT(:PPSCID, '_', :PCandID, '_', :PVL, '%'))";
```
nitpicky but should there be an _ between `:PVL` and `%` since we only allow PSCID_CANDID_VISITLABEL_* and not PSCID_CANDID_VISITLABEL*

davidblader replied:
also another not so big deal but i am curious as to why you're using `LOWER(...)` here, since value matching in columns tends to be case insensitive and most PSCIDs and VL's (at least one's i've seen) are all uppercase
sorry if this is a stupid question 😝

MounaSafiHarab replied:
I believe our MRI code some time ago was not case sensitive for the visits (IBIS has visits that are either V3 or v3); I know the DICOM archive module does it this way, perhaps for these cases? not sure

davidblader commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+        $result = array();
+        $entry  = array();
+
+        foreach ($rows as $row) {
```
i know you explain this in the docs but i think it would be useful to just briefly remind what this loop is doing
maybe at the start of the loop you can just write out an abbrieviated version of the desired structure i.e,
`[{Tarname: '...',`
  `SeriesInfo: [{...}]`
  ` },`
    ...
`]`

since it is a little bit confusing at face value

MounaSafiHarab replied:
Okay!

davidblader commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+    {
+        $factory      = \NDB_Factory::singleton();
+        $db           = $factory->Database();
+        $partial_path = $db->pselectRow(
```
could be slightly simplified with `pselectOne`

MounaSafiHarab replied:
Ok!

cmadjar commented on docs/API/LorisRESTAPI_v0.0.3.md:
````diff
@@ -728,3 +728,63 @@ The JSON object is of the form:
+GET /candidates/$CandID/$VisitLabel/dicoms/$Tarname
+```
+
+Returns a `tar` file which contains a `.meta` and a `.log` text files, 
````
OK, I found out why it was not working, the SessionID was NULL for that archive. Maybe in the query in line 104 of `visits/Dicoms.php`, there should be a join to the session table and candidate table, as it is done for the `dicoms/Dicom.php` script? Technically, we should not be able to view DICOM archives that you cannot download.

MounaSafiHarab replied:
but if your session ID is NULL< isn;t it because you ran dicomTar but not tarchiveLoader?
in that case, you would still have a DICOM that should be downloadable, non?
the path will not have the year in it, but the script does not care as it simply appends your config settings to where the .tar files are (usually in the tarchive/ folder) to the rchiveLocation in the tarchive table (with or without a year). 
To me, as long as there is metadata, this means the tarchive_series, tarchive_files have been populated, meaning there is a .tar somewhere that should be downloadable...

cmadjar replied:
I agree, we want to be able to download them but for some reason, it does not work.

Here is my archiveLocation:
![screen shot 2018-05-10 at 3 43 38 pm](https://user-images.githubusercontent.com/1402456/39890742-aff3993e-5469-11e8-9a08-6d8ff5c88c60.png)

Here is what I get when I try to download it:
![screen shot 2018-05-10 at 3 43 46 pm](https://user-images.githubusercontent.com/1402456/39890762-be7fc16c-5469-11e8-990b-be384f88e59b.png)

Weird... I don't see why this would not work... @MounaSafiHarab Am I missing something?

cmadjar commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    /**
+     * Gets the Tarname saved in the database for this file
```
Nitpicking... Technically, you get the relative location of the tarchive from the tarchive directory.

MounaSafiHarab replied:
i agree, corrected

driusan commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+ * @package  API
+ * @author   Mouna Safi-Harab <mouna.safiharab@gmail.com>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
```
phpcs insists on it with the standards we use (once for the file, once for the class). It's annoying, but true throughout LORIS..

MounaSafiHarab replied:
I think it is a PHPCS requirement? not sure though. Check all our PHP scripts in all our modules, we always have this.

cmadjar replied:
... not pretty and redundant but that's life. Good to know that so I don't make that comment for each PR :P. Thanks!

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+/**
+ * Handles API requests for the candidate's dicom files. Extends
+ * Visit so that the constructor will validate the candidate
+ * and visit_label portion of URL automatically.
```
I got it from `Images.php` originally written by Dave, so I assume since he used URL, it is properly the right thing to do here (as the only real anglophone among the 3 of us)?

driusan replied:
no, @cmadjar is right. It's a typo. It's not grammatically correct without "the".

MounaSafiHarab replied:
ok then, will add `the` :)

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+require_once '../Visit.php';
+
+/**
+ * Handles API requests for the candidate's dicom files. Extends
```
done

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * Handles API requests for the candidate's dicom files
```
ok!

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/Dicoms.php:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /**
+     * Gets a list of Tars for this visit. Tarname only.
```
done!

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+class Dicom extends \Loris\API\Candidates\Candidate\Visit
+{
+    /**
+     * Construct a Dicom class object to output the candidate visits dicom tar file
```
ok

MounaSafiHarab commented on htdocs/api/v0.0.3-dev/candidates/visits/dicoms/Dicom.php:
```diff
@@ -0,0 +1,147 @@
+     * @param string $method     The method of the HTTP request
+     * @param string $CandID     The CandID to be serialized
+     * @param string $VisitLabel The visit label to be serialized
+     * @param string $Tarname    The Tar name to be retrieved
```
ok

MounaSafiHarab commented:
> a general comment for everyone that I mentioned to Xavier last week (probably not for this PR but we should keep this in mind). If a user download data via the API and then during the course of the study a participant has become inactive or completely withdrew his consent. Would there be a way to notify the user that those files should not be used anymore in analyses and that he/she should remove those files from his server? Actually, this is true too for data that were labelled to the wrong patient and had to be removed and reinserted with the proper label.

A perfect use case for @ridz1208 's notification system?:)

MounaSafiHarab commented:
@davidblader 
I will remove the passed manual test so that it cna be tested again as I pushed a couple of commits since then...

MounaSafiHarab commented on docs/API/LorisRESTAPI_v0.0.3.md:
```diff
@@ -728,3 +728,63 @@ The JSON object is of the form:
+            "SeriesDescription" : "MPRAGE_ipat2",
+            "SeriesNumber" : "2",
+            "EchoTime" : "2.98",
+            "SeriesUID" : "1.2.3.4.1107",
```
Initially the plan was to add just the minimum needed to distinguish the unique series within the .tar (hence echo time and series number were added doe the multi echo cases). We addesd series description just to give some meaning to the series. But I don;t see any harm in adding more headers.
So I did add TR, TI, slice thickness, and modality. 
The number of files and sequence name could become problematic (check the DTI entries you have in DICOM archive, each direction is a separate SequenceName, eventhough it is technically just one series; same for the number of files)... so we really do not have a one to one correspondance...
so while I agree these 2 extra fields are useful, until we resolve this issue of one "sequence" resulting in many rows, I think it is easier for now to leave them out.

cmadjar replied:
sounds good! Thanks!

thomasbeaudry commented:
a general comment for everyone that I mentioned to Xavier last week (probably not for this PR but we should keep this in mind). If a user download data via the API and then during the course of the study a participant has become inactive or completely withdrew his consent. Would there be a way to notify the user that those files should not be used anymore in analyses and that he/she should remove those files from his server? Actually, this is true too for data that were labelled to the wrong patient and had to be removed and reinserted with the proper label.

I think once a user has downloaded the data it's too late to  monitor or take back data from the user.  The data could have already been used in their own analyses, so then how would go about that.  Make them redo / discredit their results since a user can no longer be used?  A proper study is going to use data releases for their studies and you would just

MounaSafiHarab commented:
> @MounaSafiHarab I just tested the PR. I think option 2 is probably the best one as it will be rare that a tarchive won't be associated to a session anyway and it is probably best to make sure that users don't download data from non-active candidate and session.
> 
> One additional comment. I did not mean to test that specifically but happened to do it. It looks like users that don't have permissions to the DICOM archive module can still have access to the DICOM data via the API and download it. Is this expected behavior?

sounds good.  So this is the option I have in my last commit, right?

as to your second question, you are right, the API does not have permissions check (not yet).

cmadjar commented:
Right. The behaviour from the last commit is option 2.

For the permissions, if that does not exist yet for the API, I can then pass your PR. :)

ridz1208 commented:
@driusan please dismiss/rereview

MounaSafiHarab commented:
`Content-disposdition` set to attachment as per an earlier comment that I missed.
I also made the same change to the `images/Image.php` section. 

also confirmed with @driusan that the version remains at `v0.0.3-dev`; i.e. no need for making a `v0.0.4-dev` version as `v0.0.3` not release yet.
